sdk/communication/Azure.Communication.Calling.Server/src/PartitionedDownloader.cs

```diff
@@ -0,0 +1,323 @@
+                }
+                catch (RequestFailedException ex) when (ex.ErrorCode == "Invalid Range")
+                {
+                    initialResponse = await _client.DownloadStreamingAsync(
+                        endpoint,
+                        range: default,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+                }
+
+                // If the first segment was the entire blob, we'll copy that to
+                // the output stream and finish now
+                long initialLength = ParseResponseContentLength(initialResponse);
+                long totalLength = ParseRangeTotalLength(initialResponse);
+                if (initialLength == totalLength)
+                {
+                    await CopyToAsync(
+                        initialResponse,
+                        destination,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+                    return initialResponse.GetRawResponse();
+                }
+
+                // Create a queue of tasks that will each download one segment
+                // of the blob.  The queue maintains the order of the segments
+                // so we can keep appending to the end of the destination
+                // stream when each segment finishes.
+                var runningTasks = new Queue<Task<Response<Stream>>>();
+                runningTasks.Enqueue(initialResponseTask);
+
+                // Fill the queue with tasks to download each of the remaining
+                // ranges in the blob
+                foreach (HttpRange httpRange in GetRanges(initialLength, totalLength))
+                {
+                    // Add the next Task (which will start the download but
+                    // return before it's completed downloading)
+                    runningTasks.Enqueue(_client.DownloadStreamingAsync(
+                        endpoint,
+                        httpRange,
+                        cancellationToken));
+
+                    // If we have fewer tasks than alotted workers, then just
+                    // continue adding tasks until we have _maxWorkerCount
+                    // running in parallel
+                    if (runningTasks.Count < _maxWorkerCount)
+                    {
+                        continue;
+                    }
+
+                    // Once all the workers are busy, wait for the first
+                    // segment to finish downloading before we create more work
+                    await ConsumeQueuedTask().ConfigureAwait(false);
+                }
+
+                // Wait for all of the remaining segments to download
+                while (runningTasks.Count > 0)
+                {
+                    await ConsumeQueuedTask().ConfigureAwait(false);
+                }
+
+                return initialResponse.GetRawResponse();
```

Comment:
@christothes do we have any guidelines for returning aggregated responses?

Reply:
Sorry, I missed this earlier... What does the aggregate response look like in this case?

Reply:
@christothes, what this method does is:
- Get a first range (including Response) of data.
- Calculates for more chunks of data needed and requests them.
- It stores the different chunks in a stream or a file, and ultimately it only returns the first response received.